### DISCUSSION TITLE
build nxt graph from local file structure

### DISCUSSION THREAD

**Discussion by FXTD-ODYSSEY** [nxt-dev/nxt_editor#219] [Ideas]
node code nested in the `.nxt`  file which hard to review and maintain.
I want to build a nxt_graph from local file structure.

here is a basic example what I want
```
my_nxt_graph/
├─ first_node/
│  ├─ 01@first_process_node.py
│  ├─ 02@second_process_node.py
│  ├─ __init__.py
├─ second_node/
│  ├─ 01@create_new_scene.py
│  ├─ 02@import_file.py
│  ├─ __init__.py
├─ __init__.py
```

then I can using a cli tool to build `my_nxt_graph.nxt` file

![image](https://user-images.githubusercontent.com/40897360/141037756-20d2b769-c0e9-4e8d-b64e-101ecf30e7cb.png)

How to manage the connections and attrs?
My raw idea is define in the docstring.

here is the `second_node/__init__.py` docstring example.

```python
"""
"exec_input" : "/first_node"
"comment" : "second_node"
--- nxt attrs splitter ---
bla,bla - real string data in node
"""
```

---

I have no idea how to manage the layer right now.

**Comment by MichaelAldrich** [comment]
I love this idea. There are lots of open-ended questions I think. Right now, you can work kinda like this if you use the `contents::` [token](https://nxt-dev.github.io/reference/#tokens), which can be a good half-step if you'd like to write code in regular python files that runs in a node.

Here are some questions with some of my best-guess answers at what might work well, it's an idea with lots of corners.

- Since this file structure looks like a normal python package, **what is the developer experience like to import and use code defined from nxt?** maybe:
    ```python
    from my_nxt_graph import first_process_node
    ```
- You're showing "1 python file per node", but "1 python file per layer" could also be an interesting pattern.
  - This has a connection to the "How to manage the connections and attrs?" question.
  - One version of this I've thought about is a decorator on a fucntion:
  ```python
  @nxt_node("/second_node", exec_input="/first_node")
  def second_node(path="${path::other_file.py}")
  ```
- **Should nxt be able to "open" regular python packages like graphs?** This has some appealing effects when it comes to codebose explorability and experimentation.
- **Resolve gets messy** It's probably best not to let people save something named `.py` that wouldn't be valid python syntax. Would this saving mode prevent people from using resolve? This feels like the biggest barrier to me when it comes to making this feel integrated into nxt.

**Reply by FXTD-ODYSSEY** [reply]
Thanks for your reply, `contents token` is a good solution right now.

---

```
from my_nxt_graph import first_process_node
```

nxt node code not a valid python syntax, it's pretty hard to make it works like a module.
but I have a tricky idea which borrow from [asottile/future-fstrings: A backport of fstrings to python<3.6 (github.com)](https://github.com/asottile/future-fstrings).
future-fstrings make a backport for python2 to parse special prefix `f` in front of the string.
Dive into the source code, I found that it utilize the modified [tokenize](https://docs.python.org/2.7/library/tokenize.html) module to parse this.
Since nxt node code just string, it may much more easier to find out the special marker inside the code.

---

```
@nxt_node("/second_node", exec_input="/first_node")
def second_node(path="${path::other_file.py}")
```

I like this idea, it;'s a more compact than my version.
but I wonder how to get the source string code into nxt node.
